anchore_engine/util/docker.py

```diff
@@ -149,3 +151,238 @@ def parse_dockerimage_string(instr, strict=True):
+            else:
+                cmd = None
+            if cmd:
+                dockerfile_contents = dockerfile_contents + cmd + "\n"
+
+        return dockerfile_contents
+
+
+class DockerV2ManifestMetadata:
+    """
+    Processed manifest data that exposes the history, inferred dockerfile, etc based on a Docker Image Manifest v2 schema 2: https://docs.docker.com/registry/spec/manifest-v2-2/
+    """
+
+    def __init__(self, manifest_json, image_config_json):
+        """
+
+        :param manifest_json:
+        :param image_config_json:
+        :param downloaded_blob_list: list of string filenames for blobs downloaded, assumed to be sh256 digests
+        """
+
+        self.raw = manifest_json
+        self.image_config = image_config_json
+        self.layer_ids = self._layer_ids()
+        self.history = self._history()
+        self.inferred_dockerfile = self._infer_dockerfile()
+        self.architecture = self._architecture()
+
+    def _architecture(self):
+        if not self.image_config:
+            # Cannot get it without the config, manifest is insufficient
+            return None
+
+        return self.image_config.get("architecture")
+
+    def _layer_ids(self):
+        # Filter out all empty layers
+        return [
+            x.get("digest")
+            for x in self.raw.get("layers")
+            if not x.get("empty_layer", False)
+        ]
+
+    def _history(self):
+        # add support for cases where image metadata does not contain a history element at all
+        rawhistory = copy.copy(self.image_config.get("history"))
+        rawlayers = copy.copy(
+            list(self.raw.get("layers", []))
+        )  # This is carried over from old impl
+
+        if rawhistory is None:
+            # Construct the right number of empty history elements for each layer
+            rawhistory = [{} for layer in rawlayers]
+
+        done = False
+        history = []
+        idx = 0
+        for entry in rawhistory:
+            if entry.get("empty_layer", False):
+                lid = "<missing>"
+                lsize = 0
+            else:
+                lel = rawlayers.pop(0)
+                lid = lel["digest"]
+                lsize = lel["size"]
+
+            history.append(
+                {
+                    "Created": entry.get("created", ""),
+                    "CreatedBy": entry.get("created_by", ""),
+                    "Comment": "",
+                    "Id": lid,
+                    "Size": lsize,
+                    "Tags": [],
+                }
+            )
+
+            idx += 1
+
+        return history
+
+    def _infer_dockerfile(self):
+        dockerfile_contents = "FROM scratch\n"
+        for hel in self._history():
+            patt = re.match(r"^/bin/sh -c #\(nop\) +(.*)", hel["CreatedBy"])
+            if patt:
+                cmd = patt.group(1)
+            elif hel["CreatedBy"]:
+                cmd = "RUN " + hel["CreatedBy"]
+            else:
```

Comment:
same suggestion as before, I think this can continue and then use `+=` on the assignment